apps/mason2/simulate_illumina.cpp

```diff
@@ -183,22 +186,13 @@ void _simulateSequence(TRead & read, TRng & rng, TFrag const & frag,
             continue;
         }
 
+        // NOTE: We can only insert CGAT, but we can have a polymorphism to N.
+        dist.param(param_t{0u, static_cast<unsigned short>(cigar[i].operation == 'I' ? 3u : 4u)});
+
         // Otherwise, we have insertions or mismatches.
         for (unsigned j = 0; j < cigar[i].count; ++j)
         {
-            // Pick a value between 0 and 1.
-            std::uniform_real_distribution<double> dist(0, 1);
-            double x = 1.0;
-            while (x == 1.0)
-                x = dist(rng);
-            int num = static_cast<int>(x / 0.25);
-
-            // NOTE: We can only insert CGAT, but we can have a polymorphism to N.
-
-            if (cigar[i].operation == 'I')
-                appendValue(read, seqan2::Dna5(num));
-            else
-                appendValue(read, seqan2::Dna5(num + (num == ordValue(*it))));
+            appendValue(read, seqan2::Dna5(dist(rng)));
```

Comment:
This does not guarantee that in case of a substitution the simulated symbol is different to the one in the fragment sequence, does it? So in 20% of the cases it is the same symbol. We would need to draw randomly until we get a distinct symbol.
That could potentially lead to long execution times. So instead we could always draw in the range from 0..3, and add a case distinction for the substitutions, such that all drawn ranks greater or equal than the `ordValue(*it)` are incremented by one.

Reply:
Yes, substitutions were wrong.

> We would need to draw randomly until we get a distinct symbol.
> That could potentially lead to long execution times.

This is usually fine. It's only 1 in 4 that we need to draw two times, 1 in 16 that we need to draw three times, etc.
And drawing is quite fast.
mason genome does the same, and we do the same in seqan3.

> So instead we could always draw in the range from 0..3, and add a case distinction for the substitutions, such that all drawn ranks greater or equal than the ordValue(*it) are incremented by one.

But then you would have two problems:
* You would need a modulo to handle T->A
* The distribution would be skewed, instead of 1/3 for each substitution, the next rank would have 1/2, and the two others 1/4. Similar to what we have now.

Reply:
No it should not be skewed, since you randomly picked from the set DNA\{base of fragment}. There is no modulo required. This is what I thought of:
```cpp
rank += (rank >= ordValue(*it));
```
So whenever the random rank is less than `ordValue(*it)` we can use it. Otherwise if the random rank is greater than or equal to `ordValue(*it)` we add one to get the correct base whose rank is greater than the one of `ordValue(*it)`.
Of course if you only consider dna4 then the distribution would be parameterised over the interval `[0..2]`.

Reply:
Yes, that would work. I assumed `[0,3]`.

But then I would either need two distributions or change the parameters. Is that preferable to just redrawing?

Reply:
Not if you can guarantee that it never happens that for some odd reasons redrawing stalls for a long time. Which is arguably very very very unlikely and yet can cause a lot of headache if it happens. 
So I would slightly prefer single draw over redraw.

Reply:
> Not if you can guarantee that it never happens that for some odd reasons redrawing stalls for a long time. Which is arguably very very very unlikely and yet can cause a lot of headache if it happens.

Well, it would require that something in the pseudorandom generation breaks. I suspect if that ever happens, mason stalling is the least of the problems that ensue (cryptography also relies on pseudorandom number generation, no?).

Anyway, so I just do two random distributions `dna_dist_insertion`, `dna_dist_substitution`?
Just asking because I don't want to regenerate the test files each time :)